src/main/java/io/cdap/plugin/http/sink/batch/MessageBuffer.java

```diff
@@ -0,0 +1,205 @@
+import com.google.gson.internal.bind.JsonTreeWriter;
+import com.google.gson.stream.JsonWriter;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.format.StructuredRecordStringConverter;
+import org.json4s.jackson.Json;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * MessageBuffer is used to store the structured records in a buffer till the batch size is reached.
+ * Once the batch size is reached, the records are converted to the appropriate format and appended to the message.
+ * The message is then returned to the HTTPRecordWriter.
+ */
+public class MessageBuffer {
+  private static final String REGEX_HASHED_VAR = "#s*(\\w+)";
+  private final List<StructuredRecord> buffer;
+  private final String jsonBatchKey;
+  private final Boolean shouldWriteJsonAsArray;
+  private final String delimiterForMessages;
+  private final String charset;
+  private final String customMessageBody;
+  private final Function<List<StructuredRecord>, String> messageFormatter;
+  private final String contentType;
+
+  public MessageBuffer(
+          String messageFormat, String jsonBatchKey, boolean shouldWriteJsonAsArray,
+          String delimiterForMessages, String charset, String customMessageBody
+  ) {
+    this.jsonBatchKey = jsonBatchKey;
+    this.delimiterForMessages = delimiterForMessages;
+    this.charset = charset;
+    this.shouldWriteJsonAsArray = shouldWriteJsonAsArray;
+    this.customMessageBody = customMessageBody;
+    this.buffer = new ArrayList<>();
+    switch (messageFormat.toLowerCase()) {
+      case "json":
+        messageFormatter = this::formatAsJsonNoException;
+        contentType = "application/json";
+        break;
+      case "form":
+        messageFormatter = this::formatAsForm;
+        contentType = "application/x-www-form-urlencoded";
+        break;
+      case "custom":
+        messageFormatter = this::formatAsCustom;
+        contentType = "text/plain";
+        break;
+      default:
+        throw new IllegalArgumentException("Invalid message format: " + messageFormat);
+    }
+  }
+
+  public void add(StructuredRecord record) {
+    buffer.add(record);
+  }
+
+  public void clear() {
+    buffer.clear();
+  }
+
+  public int size() {
+    return buffer.size();
+  }
+
+  public boolean isEmpty() {
+    return buffer.isEmpty();
+  }
+
+  public String getContentType() {
+    return contentType;
+  }
+
+  public String getMessage() throws IOException {
+    return messageFormatter.apply(buffer);
+  }
+
+  private String formatAsJsonNoException(List<StructuredRecord> buffer) {
```

Comment:
The method name confuses the purpose of the method. Ideally the method name that is to be used throughout the class should hide the implementation details, so I'd suggest changing this method to `formatAsJson()` and renaming the other one to something else.

Reply:
Renamed to `formatAsJson` and the other to `formatAsJsonInternal`